src/utilities/filetypes/AlfalfaJSON.cpp

```diff
@@ -0,0 +1,223 @@
+    IddObjectType idd_type = idf_object.iddObject().type();
+    std::string _display_name = display_name;
+
+    if (display_name.size() == 0) {
+      boost::optional<std::string> name = getName(idf_object);
+      if (name.is_initialized()) {
+        _display_name = name.get();
+      }
+    }
+
+    AlfalfaComponent* component;
+    try {
+      switch (idd_type.value()) {
+        case IddObjectType::OS_Output_Meter:
+        case IddObjectType::Output_Meter:
+          component = new AlfalfaMeter(idf_object);
+          break;
+        case IddObjectType::OS_EnergyManagementSystem_Actuator:
+        case IddObjectType::EnergyManagementSystem_Actuator:
+          component = new AlfalfaActuator(idf_object);
+          break;
+        case IddObjectType::OS_Output_Variable:
+        case IddObjectType::Output_Variable:
+        case IddObjectType::EnergyManagementSystem_OutputVariable:
+        case IddObjectType::OS_EnergyManagementSystem_OutputVariable:
+          component = new AlfalfaOutputVariable(idf_object);
+          break;
+        case IddObjectType::OS_EnergyManagementSystem_GlobalVariable:
+        case IddObjectType::EnergyManagementSystem_GlobalVariable:
+          component = new AlfalfaGlobalVariable(idf_object);
+          break;
+        default:
+          LOG(Error, "Unable to create Alfalfa Point from Object of type " + idd_type.valueDescription());
+          return boost::none;
+      }
+    } catch (std::runtime_error& error) {
+      LOG(Error, error.what());
+      return boost::none;
+    }
+    return exposeFromComponent(*component, _display_name);
+  }
+
+  boost::optional<AlfalfaPoint> AlfalfaJSON::exposeFromComponent(const AlfalfaComponent& component, const std::string& display_name) {
+    std::string _display_name = display_name;
+    if (display_name.size() == 0) {
+      if (component.type == "Actuator") {
+        _display_name = "Actuator for " + component.parameters["component_name"].asString() + ":" + component.parameters["component_type"].asString()
+                        + ":" + component.parameters["control_type"].asString();
+      } else if (component.type == "Constant") {
+        LOG(Error, "Constant points must be provided with a display name");
+        return boost::none;
+      } else if (component.type == "Meter") {
+        _display_name = "Meter for " + component.parameters["meter_name"].asString();
+      } else if (component.type == "GlobalVariable") {
+        _display_name = "Global Variable for " + component.parameters["variable_name"].asString();
+      } else if (component.type == "OutputVariable") {
+        _display_name =
+          "Output Variable for " + component.parameters["variable_key"].asString() + ":" + component.parameters["variable_name"].asString();
+      } else {
+        LOG(Error, "Invalid component type");
+        return boost::none;
+      }
+    }
```

Comment:
Kinda similar. I guess I'm questionning whether this logic should instead be inside the Derived class of AlfalfaComponent.

AlfalfaComponent would have a `virtual Json::Value exposeToAlfalfa() const = 0` method or whatever the name (maybe just toJSON? or alfalfaParameters?), and each derived class would implement the business logic where it's most logical to place it? Making it a pure virtual method would make AlfalfaComponent an abstract class, which I guess is what it should be.

Reply:
If you look through the git commit history I originally had everything going off a base class with fully virtual functions. The problem was I couldn't access those in functions that just took a reference to the base class so any child type could be passed in.